_drafts/2024-05-18-get-started-with-aws-tf-01.md

```diff
@@ -0,0 +1,254 @@
+
+1. Have an [AWS Account](https://aws.amazon.com) setup
+2. Installed [Terraform cli](https://developer.hashicorp.com/terraform/install?product_intent=terraform)
+
+## Step 0 - Create a folder for your project
+
+As we'll be building our infractructure using Terraform, which is an IaC tool, we'll need somewhere to write this code.
+Let's create a folder called `my-unicorn-project` and inside that folder create another folder called `infra`.
+
+## Step 1 - Define our Terraform provider
+
+Now we will create our `main.tf` file inside our `infra` folder where we will define our cloud provider.
+
+```terraform
+# /infra/main.tf
+
+terraform {
+  required_version = ">= 1.2.0"
+  required_providers {
+    aws = {
+            source = "hashicorp/aws"
+            version = ">= 5.41.0"
+        }
+    }
+}
+```
+
+Now when you open a terminal in the current directory and run `terraform init` Terraform will download the required provider modules so they can be used to configure our services.
+
+
+## Step 2 - Create an AWS EC2 instance
+
+Now we have successfully initialized terraform with the aws provider we can define aws resources in our code. In this step we will create two things:
+1. An AWS EC2 instance resource
+2. An aws ami image data reference for EC2 to run
+
+Create a new file called `unicorn-api.tf`.  
+Inside the file, add:
+
+```terraform
+# /infra/unicorn-api.tf
+
+data "aws_ami" "amazon_x86_image" {
+    most_recent = true
+    owners = ["amazon"]
+    name_regex = "^al2023-ami-ecs.*"
+
+    filter {
+        name = "architecure"
+        values = ["x86_64"]
+    }
+}
+
+resource "aws_instance" "unicorn_vm" {
+    ami = data.aws_ami.amazon_x86_image.id
+    instance_type = "t2.micro"
+
+    tags = {
+        Name = "Unicorn"
+    }
+}
+```
+
+Now when you open the terminal and run `terraform plan` the logs will show that we're adding one new resource.  
+
+Now try running `terraform apply` to... err... apply the plan.  
+
+It fails! But why?
+
+Even though we've created an aws account, and downloaded Terraform, and the aws provider for Terraform to use, we still need Terraform to authenticate with aws in order to create these resources.
+
+## Step 2.1 - Create an AWS API Key for terraform
+
+In order for Terraform to manage resources in AWS it needs access to an aws account. We can accomplish this by creating an aws account specifically for this tool and defining corresponding API keys for it.
+
+Create a new file called `aws-provider-key.tf` and add the following content:
+
+```terraform
+# /infra/aws-provider-key.tf
+
+provider "aws" {
+    access_key = ""
+    secret_key = ""
+    region = "ap-southeast-2"
+}
+```
+
+You can opt to use a region closer to you. As I'm in Australia I'm opting to use the australian region.
+
+Next you need to create two things in the AWS Management Console:
+1. Create a [new User](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html) with administrator permission
```

Comment:
This is a pet peeve of mine - instructional blog posts telling people to provide full access to a set of AWS credentials. Ideally we should be demonstrating and promoting good security practices and suggest using a more limited policy. There's two ways to address this (imo) with varying levels of work required by you:

- Continue with full admin permissions but very clearly highlight that giving the API keys full administrative access is a bad idea, quantify the risk and heavily signpost that you shouldn't make a habit of this
- Figure out an appropriate IAM policy that would allow access only to the operations needed to complete this tutorial and provide that policy in your post.

The latter is obviously a lot more work for you compared to the former but might be a useful learning exercise for you.